marimo/_runtime/executor.py

```diff
@@ -0,0 +1,389 @@
+            ]
+            if key in glbls
+        }
+
+        from asyncio import Future
+
+        from marimo._plugins.ui._core.ui_element import UIElement
+        from marimo._runtime.state import SetFunctor, State
+
+        for ref in refs:
+            if ref in glbls:
+                if (
+                    isinstance(
+                        glbls[ref],
+                        (State, SetFunctor, ZeroCopy, Future, UIElement),
+                    )
+                    or inspect.ismodule(glbls[ref])
+                    or inspect.isfunction(glbls[ref])
+                    or from_unclonable_module(glbls[ref])
+                    or is_unclonable_type(glbls[ref])
+                ):
+                    lcls[ref] = glbls[ref]
+                elif isinstance(glbls[ref], ShallowCopy):
+                    lcls[ref] = shallow_copy(glbls[ref])
+                else:
+                    try:
+                        lcls[ref] = deepcopy(glbls[ref])
+                    except TypeError as e:
+                        raise CloneError(
+                            f"Could not clone reference `{ref}` of type "
+                            f"{getattr(glbls[ref], '__module__', '<module>')}."
+                            f"{glbls[ref].__class__.__name__}"
+                            " try wrapping the object in a `zero_copy`"
+                            "call. If this is a common object type, consider "
+                            "making an issue on the marimo GitHub "
+                            "repository to never deepcopy."
+                        ) from e
+            elif ref not in glbls["__builtins__"]:
+                if ref in cell.defs:
+                    raise MarimoNameError(
+                        f"name `{ref}` is referenced before definition."
+                    )
+                raise MarimoMissingRefError(ref)
+
+        # NOTE: Execution expects the globals dictionary by memory reference,
+        # so we need to clear it and update it with the sanitized locals,
+        # returning a backup of the original globals for later restoration.
+        # This must be performed at the end of the function to ensure valid
+        # state in case of failure.
+        backup = {**glbls}
+        glbls.clear()
+        glbls.update(lcls)
+        return backup
+
+    @staticmethod
+    def update_outputs(
+        cell: CellImpl, glbls: dict[str, Any], backup: dict[str, Any]
+    ) -> None:
+        # NOTE: After execution, restore global state and update outputs.
+        lcls = {**glbls}
+        glbls.clear()
+        glbls.update(backup)
+
+        defs = cell.defs
+        for df in defs:
+            if df in lcls:
+                # Overwrite will delete the reference.
+                # Weak copy holds on with references.
+                glbls[df] = lcls[df]
+            elif df in glbls:
+                del glbls[df]
+
+        for df in backup:
+            if df.startswith(f"_cell_{cell.cell_id}_"):
```

Comment:
Should this be moved to a utility function? There's an `is_local` function in `visitor`, for example.

Reply:
Moved, but in retrospect maybe `marimo.ast.variables` might have been better and moved over the data classes too.